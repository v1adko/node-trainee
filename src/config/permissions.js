export default {
  PUBLIC: { role: 'public', value: 0 },
  USER: { role: 'user', value: 1 },
  ADMIN: { role: 'admin', value: 2 }
};
