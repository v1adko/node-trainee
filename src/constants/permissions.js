export default {
  PUBLIC: { value: 'public', priority: 0 },
  USER: { value: 'user', priority: 1 },
  MODERATOR: { value: 'moderator', priority: 2 },
  ADMIN: { value: 'admin', priority: 3 }
};
