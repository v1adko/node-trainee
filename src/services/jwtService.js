import jwt from 'jsonwebtoken';
import configJwt from '../config/jwt';

const { tokenSecret: secret, tokenExpiresIn: expiresIn } = configJwt;

const options = { expiresIn };

class JwtService {
  generateJwt = (user) => {
    const payload = {
      _id: user._id,
      role: user.role
    };
    return jwt.sign(payload, secret, options);
  };

  decoder = (token) => {
    const promise = new Promise((resolve, reject) => {
      jwt.verify(token, secret, (error, decoded) => {
        if (error) {
          reject(new Error('Failed to authenticate token.'));
        } else {
          resolve(decoded);
        }
      });
    });
    return promise;
  };
}

export default new JwtService();
