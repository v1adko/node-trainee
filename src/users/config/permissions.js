export default {
  CREATE_USER: 'createUser',
  READ_USER: 'readUser',
  UPDATE_USER: 'updateUser',
  DELETE_USER: 'deleteUser',
  GET_MY_DATE: 'getMyDate',
  UPDATE_MY_DATE: 'updateMyDate'
};
