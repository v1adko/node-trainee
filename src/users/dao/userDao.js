const BaseDao = require('./baseDaoMongoose');
const User = require('../models/user');

class UserDao extends BaseDao {}

module.exports = new UserDao(User);
