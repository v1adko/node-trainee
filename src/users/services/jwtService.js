const jwt = require('jsonwebtoken');

const { secretTokkenWord: secret } = require('../config/jwt');

const permissionsConst = require('../config/permissions');

class JwtService {
  constructor() {
    this.generateJwt = (user) => {
      const payload = {
        _id: user._id,
        permissions: Object.values(permissionsConst)
      };
      const options = {
        expiresIn: '2h'
      };

      return jwt.sign(payload, secret, options);
    };
  }
}

module.exports = new JwtService();
