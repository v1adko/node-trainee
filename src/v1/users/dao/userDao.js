import BaseDao from '../../../utils/baseDaoMongoose';
import User from '../models/user';

class UserDao extends BaseDao {
  create(username, password) {
    const user = new this.Model();
    user.username = username;
    user.password = password;
    return super.create(user);
  }
}

export default new UserDao(User);
