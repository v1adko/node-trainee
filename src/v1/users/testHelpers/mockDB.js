import userDao from '../dao';
import userMocks from './userMocks';
import permissions from '../../../constants/permissions';

const { getUsersFields, user: userFields, admin: adminFields } = userMocks;

class MockDB {
  constructor(DAO) {
    this.DAO = DAO;
  }

  createDefaultUsers = async (userCount = 3) => {
    const users = [];
    getUsersFields(userCount).forEach(async (element) => {
      const { username, password, role } = element;
      const user = await this.createUser(username, password, role);
      users.push(user);
    });
  };

  createDefaultUser = async () => {
    const { username, password, role } = userFields;
    const user = await this.createUser(username, password, role);
    return user;
  };

  createDefaultAdmin = async () => {
    const { username, password, role } = adminFields;
    const user = await this.createUser(username, password, role);
    return user;
  };

  createUser = async (username, password, role = permissions.USER.value) => {
    const createdUser = this.DAO.create(username, password, role);
    return createdUser;
  };

  getAll = async () => {
    const users = await this.DAO.getAll();
    return users;
  };

  cleanDB = async () => {
    await this.DAO.Model.remove();
  };
}

export default new MockDB(userDao);
