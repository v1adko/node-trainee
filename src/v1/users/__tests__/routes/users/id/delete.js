import simulate from '../../../../../../utils/tests-utils/request-helper';
import jwtService from '../../../../../services/jwt-service';
import User from '../../../../user-model';
import userDao from '../../../../user-dao';
import mockDB from '../../../../../../utils/tests-utils/mock-db';
import permissions from '../../../../../../constants/permissions';
import UserFields from '../../../../../../utils/tests-utils/test-user-fields';

const ROUTE = '/v1/users';

const {
  username, password, invalidUserId, invalidToken
} = UserFields;
let user;
let userToken;
let adminToken;

async function clean() {
  await mockDB.cleanDB();
  userToken = null;
}

async function createUserAndAccessTokens() {
  const userActor = new User();
  const adminActor = new User();
  adminActor.role = permissions.ADMIN.value;
  userToken = jwtService.generateJwt(userActor);
  adminToken = jwtService.generateJwt(adminActor);
  user = await mockDB.createUser(username, password);
}

beforeAll(clean);
afterAll(mockDB.closeConnection);

describe(`Test the ${ROUTE}/:id path`, () => {
  afterEach(clean);
  beforeEach(createUserAndAccessTokens);

  it('should delete user by id', async () => {
    const route = `${ROUTE}/${user.id}`;
    const { status, message } = await simulate.delete(
      route,
      200,
      {},
      adminToken
    );
    const users = await userDao.getAll();

    expect(status).toBe(true);
    expect(message).toMatchSnapshot();
    expect(users).toEqual([]);
  });

  it('should not delete user, because user token is not valid', async () => {
    const route = `${ROUTE}/${user.id}`;
    const { error } = await simulate.delete(route, 401, {}, invalidToken);
    const nonDelentedUser = await userDao.getById(user.id);

    expect(nonDelentedUser.id).toEqual(user.id);
    expect(error).toMatchSnapshot();
  });

  it('should not delete user, because user token is not have enough permissions', async () => {
    const route = `${ROUTE}/${user.id}`;
    const { error } = await simulate.delete(route, 403, {}, userToken);
    const nonDelentedUser = await userDao.getById(user.id);

    expect(nonDelentedUser.id).toEqual(user.id);
    expect(error).toMatchSnapshot();
  });

  it('should not delete user, because user id is invalid', async () => {
    const route = `${ROUTE}/${invalidUserId}`;
    const { error } = await simulate.delete(route, 400, {}, adminToken);
    const nonDelentedUser = await userDao.getById(user.id);

    expect(nonDelentedUser.id).toEqual(user.id);
    expect(error).toMatchSnapshot();
  });

  it('should not delete user, bacuse user does not exist', async () => {
    await user.remove(user.id);
    const route = `${ROUTE}/${user.id}`;
    const error = await simulate.delete(route, 200, {}, adminToken);
    console.log(error);
  });
});
