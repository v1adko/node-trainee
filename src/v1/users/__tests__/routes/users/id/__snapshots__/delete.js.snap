// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the /v1/users/:id path should delete user by id 1`] = `"User was deleted"`;

exports[`Test the /v1/users/:id path should not delete user, because user id is invalid 1`] = `
Object {
  "code": 400,
  "message": "RequestValidationError(400): \\"id\\" must only contain hexadecimal characters",
  "name": "RequestValidationError",
}
`;

exports[`Test the /v1/users/:id path should not delete user, because user token is not have enough permissions 1`] = `
Object {
  "code": 403,
  "message": "PermissionsError(403): Access was denied. Not enough permissions.",
  "name": "PermissionsError",
}
`;

exports[`Test the /v1/users/:id path should not delete user, because user token is not valid 1`] = `
Object {
  "code": 401,
  "message": "AuthorizationError(401): Invalid token, please repeat authentication.",
  "name": "AuthorizationError",
}
`;
